2019/04/fourth_day.rb

```diff
@@ -1,11 +1,30 @@
-input = 353096..843212
-
-# find out, how many strings from 353096-843212 are potential passwords
-# Criteria:
-# 6 digits (all of them)
-# two adjacent digits are the same
-# digits never decrease, but increase or stay the sam
-#
-# plan:
-# iterate over all possible input-strings
-# check for criteria. if one fails: next string
+
+def is_number_potential_password?(number)
+  # has six digits
+  has_six_digits = number.to_s.length == 6
+  # two adjacent digits are the same
+  two_adjacent_digits = false
+  number_string = number.to_s
+  if number_string[0] == number_string[1] || number_string[1] == number_string[2] || number_string[2] == number_string[3] || number_string[3] == number_string[4] || number_string[4] == number_string[5]
+    two_adjacent_digits = true
+  end
+  # digits never decrease. Sort and compare if strings are the same
+  never_decrease = number.to_s.chars.sort == number.to_s.chars
+
+  return (has_six_digits && never_decrease && two_adjacent_digits)
+end
+
+def first_puzzle
+
+  potential_passwords = []
+
+  (353096..843212).each { |current_number|
```

Comment:
Instead of `each` and then appending to the array, use `map`, which returns an array. The block should return the input number, if it's a potential password, and nil otherwise. Then put a `.compact` on the resulting array.